netcreateorg/netcreate-itest
# Infrastructure: Module Architecture Stage 1

NOTE: This does not affect the operation of NetCreate ITEST at all. This codebase written to be independent of it until it's needed.

## OVERVIEW

This pull request (PR) adds the prototype framework for developing a more modular style of coding. We've designed it to accommodate future codebases as much as we can predict, but **now** we will now use it in solving **NetCreate 2.0** challenges, refining the framework as we go.

**There are no functional changes in this PR that are visible to the research team.** Most of the new code is running independently via a different command interface. 

## TESTING

This PR includes the most recent `dev` branch changes as of August 21, 2023. To test this merge request, we want to make sure there are no new errors or regressions. 

To test:

* [ ] Pull this branch
* [ ] Make sure to `nvm use` to load the correct version of node
* [ ] Remove the `node_modules` directory then `npm ci`
* [ ] Run NetCreate in the way that you have been. Everything should still work.

TECH NOTE: If you are running an **ARM Macintosh** (M1, M2, etc) and see errors, you may have not run `nvm use` to load the supported version of NodeJS (minimum version is now v14), or there is an incompatibility with the extremely out-of-date libraries currently in NetCreate. You may need to run from a terminal window that has Rosetta (Intel CPU emulation) enabled. The server startup should detect the architecture mismatch and print a warning on the terminal console. This version also now _FORCES_ i386 Intel emulation in all VSCode terminals. 

## NEXT STEPS

Now that this is in place, we need to do a major format cleanup as the many non-conforming changes made when _Prettier_ was disabled has made it difficult to work with the code. This should be relatively quick, and will be performed as a different PR after this one is accepted and merged.

Following that, we will update outdated packages as much as we can. This may involve updating the build system as well because it is out of date and no longer supported. While we could ignore this, the NetCreate 2.0 requirement for secure data connections means we should not, as out-of-date libraries are a security risk.

---
_the following describes the technical goal setting and reasoning behind this PR, and is of interest primarily to the developers_

## APPROACH

We're creating a system that will apply to future systems other than NetCreate, which requires formalizing the common operations in an app. The system core is in the `_ur` directory and is invoked using the corresponding `ur` command in the terminal. It is independent from NetCreate at this time to not affect the current software, but is designed for simple `import` into the codebase. 

There are several areas we are formalizing the common operations

#### Declarations

In the `_ur/_sys` directory are a number of files that begin with `declare-`, which are modules that model asynchronous programming and data formats. They currently serve largely as 
a reference for the concepts, but could eventually be used for runtime type validation.

#### Calling Convention Standards

There are almost a dozen different ways to invoke code in NetCreate, and there are specific contexts where one way is preferred over another. Our new framework seeks to simplify this by providing clear patterns to follow.

#### Single Module Providing Direct API, Process API, EventEmitter API, and Stream API

Write the workhorse code once, and use the same or similar API to handle four different contexts for invoking that code without having to duplicate or "wrap" it. We're trying to make "single source of truth" as easy to create and "reason with" in all the contexts that code can be used.

## EMERGING SUBSYSTEMS

There are several common APIs that are also being defined in tandem:

* "UR DataExchange" interface for serializing data (PTrack, config files, message traffic, display lists, etc) to send between various code modules anywhere on the network

* "UR Async Communication" interface to standardize the appropriate use and context for the six different ways you can asynchronously invoke code. 

* "UR Collection" interface that standardize the common operations typically performed on any datastore: persisted data in a database or file, runtime data, derived runtime data, arrays, dictionaries, maps, registries, resource pools, etc. 


## APPENDIX: JUSTIFICATION OF WORK

We need a robust modular structure to be able to secure and port the system to remote hosts. While we could hack it again to make this work, the disorganization and internal inconsistency of the code base makes every new feature added more likely to break something while taking a very long time to implement. By expending time now to clean this all up into clear concepts, we build a better base for robust and secure deployment while also improving the speed of implementation of new features. This also gives us an entry point into writing **automatic testing code**; previous codebases have required manual testing with scant resources.

We're addressing these desires in the current ITEST work by iterating on our URSYS framework (NC-2018 uses UNISYS, its precursor). On reviewing the NetCreate codebase, we noted these general issues:

* not having clear patterns for **following dataflow/reactive principles**
* not having clear patterns for **separating concerns** so they can run as independent modules free of other dependencies, and **what** those concerns were.
* not following existing patterns for **existing concerns** and their **nomenclature**
* not having patterns for implementing standard **computer science** principles

To be fair to ourselves, the Javascript world itself has struggled with this as well but now we are able to express our own **strong opinions** on how we'd like to address them, given our more nuanced understanding of our preferences and the requirements of an active research team conducting real-world operations.
